### ISSUE TITLE
CPU usage extremely high, on close memory usage skyrockets

### ISSUE THREAD

**Issue by erichiller** [Jason-Rev/vscode-spell-checker#77]
I love this extension, but I had to uninstall because it was causing some performance issues.
When running, one vscode process was taking 12% of the CPU constantly, and I traced it back here. Then, when I closed all vscode windows, the process would not close and memory usage skyrockets, at the rate of 10s of MB or RAM per second, approaching infinity. 
Here is a dump of the process from task manager:
```
Name	PID	Status	User name	CPU	Memory (private working set)	Description	Command line
Code.exe	18160	Running	ehiller	12 	408 K	Visual Studio Code	"
C:\Program Files (x86)\Microsoft VS Code\Code.exe" 
C:\Users\ehiller\.vscode\extensions\streetsidesoftware.code-spell-checker-0.17.2\node_modules\vscode-languageclient\lib\utils\electronForkStart C:\Users\ehiller\.vscode\extensions\streetsidesoftware.code-spell-checker-0.17.2\server\src\server.js --node-ipc
```

**Comment by Jason3S** [comment]
Can you install the latest version and see if it still happens?

**Comment by Jason3S** [comment]
Thank you for the report by the way.

A few questions:
Does it happen all the time or just for a single project?

Any reproduction steps you have would be useful.

**Comment by erichiller** [comment]
I have been watching this over the past few days. So far (since upgrading) I have not had the issue occur. The computer I had the issue with most, I'll be seeing again on Monday - but so far so good. Thanks!

**Comment by Jason3S** [comment]
It might be an issue that happens more often on Windows.  This issue can also happen if VS Code quits and the extension doesn't get the notification.  I have only seen high CPU usage if VS Code was forced to quit, leaving the extension running.

**Comment by erichiller** [comment]
It occurred again yesterday, I had ~7 windows and ~50 tabs running. I went through eliminating one by one. It was a `main.go` tab. In the past it has seemed to be `main.go` as well, but I can not 100% confirm this. Maybe something to look at - just wanted to provide that datapoint.

**Comment by Jason3S** [comment]
Go? Ok, thank you.  Are you running Python? or its extension?

**Comment by erichiller** [comment]
I almost said no, because I essentially never do Python work, but I checked before I spoke and it is. Here is [my config](https://gist.github.com/erichiller/3746adb251854c614f215ed87ed60980) , might also help in other ways.

In that most recent version, I have spell removed as an extension(just to avoid CPU load), but all the settings are intact, so you can see them. [Here is a version with the extension](https://gist.github.com/erichiller/3746adb251854c614f215ed87ed60980/a3eb86cb6ce1427a338fae3110d93610139e55fa) , historical, but the same.

I've noticed certainly that it seems to me an issue most often when working in [GO](https://github.com/Microsoft/vscode-go) ; possibly something related to it's continual lint-ing?

**Comment by Jason3S** [comment]
Thank you.  I had only seen you behavior you describe when working with the Python extension turned on.  When I removed it, the issue went away.  I don't have a good repro step yet.  Which makes this difficult to debug.

I'll be doing some more in GO soon, so may I can git it to happen there.

**Comment by mhrabovcin** [comment]
Same issue happening on my side. Detailed report - https://github.com/Microsoft/vscode/issues/24802#issuecomment-294751573

I am working on a project with multiple languages but mostly `python`.

**Comment by Jason3S** [comment]
I would like to fix this. But I need to consistent way to reproduce it.

I know this can be very frustrating for those of you that this is happening to. Please bear with me, I use VS Code and the spell checker everyday. I am not seeing this issue. I believe it is a real issue, because I had seen something like it a few months ago.

If you are seeing this issue consistently, please work with me to find a way to reproduce it. We will need to do a bit of sleuthing.

Questions:
* What extensions do you have installed? This is important to know, because this issue is most likely caused by the interaction between multiple extensions.
* What language are you using? Better yet, if you can point me to a github repo that I can clone, this would be ideal.
* When does it happen? On VS Code startup? Only after a while? etc.
* What OS and version are you using? If you are supper advanced, a vagrant box!

Please realize I don't have a lot of time to work on this.

**Comment by mhrabovcin** [comment]
Hey @Jason3S,

thank you for all your work you're putting into this extension!

I'll try to answer your questions which will hopefully help you to investigate this issue.

### Environment

```
$ code --list-extensions --show-versions
PeterJausovec.vscode-docker@0.0.13
bungcip.better-toml@0.2.0
donjayamanne.python@0.6.3
dzannotti.theme-spacemacs@0.0.8
gccfeli.vscode-lua@0.1.2
marcostazi.VS-code-vagrantfile@0.0.4
shanoor.vscode-nginx@0.3.1
shardulm94.trailing-spaces@0.2.11
streetsidesoftware.code-spell-checker@1.1.0
vscodevim.vim@0.6.18
```

OS: `macOS Sierra 10.12.4`. 
VSC: `1.11.2` build `6eaebe3b9c70406d67c97779468c324a7a95db0e`

Sometimes the code is executed in context of `virtualbox` machine from `dcos-docker` project or `docker` container via `docker-machine`.

### Workspace

I usually work on workspace that contains multiple repos which some of them are private so I can't share them all. Here are some of the public repos that are in my workspace:

https://github.com/dcos/dcos
https://github.com/dcos/dcos-docker
https://github.com/dcos/dcos-cli
https://github.com/mesosphere/marathon-lb
https://github.com/mesosphere/vault

Total workspace stats gathered by running `cloc`:

```
   27323 text files.
   15408 unique files.
Complex regular subexpression recursion limit (32766) exceeded at /usr/local/Cellar/cloc/1.72/libexec/bin/cloc line 7573.
Complex regular subexpression recursion limit (32766) exceeded at /usr/local/Cellar/cloc/1.72/libexec/bin/cloc line 7573.
Complex regular subexpression recursion limit (32766) exceeded at /usr/local/Cellar/cloc/1.72/libexec/bin/cloc line 7573.
Complex regular subexpression recursion limit (32766) exceeded at /usr/local/Cellar/cloc/1.72/libexec/bin/cloc line 7573.
Complex regular subexpression recursion limit (32766) exceeded at /usr/local/Cellar/cloc/1.72/libexec/bin/cloc line 7573.
Complex regular subexpression recursion limit (32766) exceeded at /usr/local/Cellar/cloc/1.72/libexec/bin/cloc line 7573.
Complex regular subexpression recursion limit (32766) exceeded at /usr/local/Cellar/cloc/1.72/libexec/bin/cloc line 7573.
Complex regular subexpression recursion limit (32766) exceeded at /usr/local/Cellar/cloc/1.72/libexec/bin/cloc line 7573.
   15726 files ignored.

github.com/AlDanial/cloc v 1.72  T=211.95 s (57.2 files/s, 19911.7 lines/s)
----------------------------------------------------------------------------------------
Language                              files          blank        comment           code
----------------------------------------------------------------------------------------
Python                                 5491         273971         352728        1168056
JSON                                   1748            527              0         530769
Go                                     1959          66898          68247         473597
C                                       500          64393         111992         440664
C/C++ Header                            341          12905          19508         109352
JavaScript                              141          22418          37656         107776
Bourne Shell                            141           8408           8802          50624
HTML                                    124           2387             58          43629
Markdown                                241           6970              0          23519
Mustache                                267            147             18          22687
PO File                                  42           8274           9834          20370
Assembly                                 92           4309           7451          16717
Java                                    203           2758           3637          16587
m4                                       20           1657            192          16203
CSS                                      51           1316            966          13173
YAML                                     65            156             49           7802
TeX                                       7            838           3594           7224
Lua                                      38           1557           1291           6689
XML                                     158            247            297           5080
make                                     52            623            373           3415
C++                                       6            650            272           2821
XSD                                      12            233             40           2414
Bourne Again Shell                      112            572            268           2347
Sass                                     19            397            128           1969
WiX source                               49            150             38           1654
DOS Batch                                34            285            108           1251
Dockerfile                               49            262            251           1249
Windows Module Definition                 6             25             14           1176
ERB                                      24            154              0           1136
Velocity Template Language               14             53             20            948
Perl                                      9            153            139            907
Cython                                    3            168             71            648
Objective C                               7             98             61            635
MSBuild script                           26             40              4            622
Expect                                    5            104            161            505
Lisp                                      1            109             81            502
INI                                      13             85             27            400
Windows Resource File                     9             54             34            301
Groovy                                   10             61             19            298
PowerShell                                3             68              6            257
Visual Basic                              4             24              1            217
WiX string localization                  10             23              0            174
C Shell                                   9             93             63            173
diff                                      1             17             87            105
Maven                                     2              5              0             94
Ruby                                      3             17              5             91
Pascal                                    1             25              0             85
SQL                                       4              3              0             42
Prolog                                    1              0              0             24
builder                                   1              1              0             18
Rust                                      1              0              0              3
----------------------------------------------------------------------------------------
SUM:                                  12129         484688         628591        3106999
----------------------------------------------------------------------------------------
```

I usually have files opened from one or two repos when working on some feature. I can reproduce CPU spikes when I am working on `python` code.

### Behaviour

When I start `code` there are is no significant CPU activity and spellchecker extension process is running with ~0 CPU usage.

I can reproduce the issue when I open a `python` file from one of the private repos. Process starts using 100% CPU instantly.

This is the `dtruss` (strace) from the time when I open the file.

```
kevent(0x7, 0x7FFF5E3B90F0, 0x0)                 = 0 0
kevent(0x7, 0x7FFF5E3B90F0, 0x0)                 = 0 0
kevent(0x7, 0x7FFF5E3B90F0, 0x0)                 = 1 0
recvmsg(0x3, 0x7FFF5E3B8EA8, 0x0)                = 8192 0
kevent(0x7, 0x7FFF5E3B90F0, 0x0)                 = 1 0
recvmsg(0x3, 0x7FFF5E3B8EA8, 0x0)                = 8192 0
kevent(0x7, 0x7FFF5E3B90F0, 0x0)                 = 1 0
recvmsg(0x3, 0x7FFF5E3B8EA8, 0x0)                = 6139 0
mmap(0x1AA4DAF00000, 0x100000, 0x3, 0x1012, 0xFF000000, 0x0)             = 0x1AA4DAF00000 0
psynch_cvsignal(0x7FDFE0D0CAE8, 0x2C5000002C700, 0x2C500)                = 256 0
psynch_cvwait(0x7FDFE0D0CAE8, 0x2C4000002C600, 0x2C400)          = 0 0
mmap(0x36CE5400000, 0x100000, 0x0, 0x1052, 0xFF000000, 0x0)              = 0x36CE5400000 0
dtrace: error on enabled probe ID 2132 (ID 320: syscall::write:return): invalid kernel access in action #12 at DIF offset 92
dtrace: error on enabled probe ID 2132 (ID 320: syscall::write:return): invalid kernel access in action #12 at DIF offset 92
dtrace: error on enabled probe ID 2132 (ID 320: syscall::write:return): invalid kernel access in action #12 at DIF offset 92
kevent(0x7, 0x7FFF5E3B90F0, 0x0)                 = 0 0
kevent(0x7, 0x7FFF5E3B90F0, 0x0)                 = 0 0
psynch_cvsignal(0x7FDFE0D0CAE8, 0x2C6000002C800, 0x2C600)                = 256 0
psynch_cvwait(0x7FDFE0D0CAE8, 0x2C5000002C700, 0x2C500)          = 0 0
psynch_cvsignal(0x7FDFE0D0CAE8, 0x2C7000002C900, 0x2C700)                = 256 0
psynch_cvwait(0x7FDFE0D0CAE8, 0x2C6000002C800, 0x2C600)          = 0 0
madvise(0x109E8C000, 0x46000, 0x9)               = 0 0
madvise(0x106DC1000, 0x22000, 0x9)               = 0 0
thread_selfid(0x7000013B4000, 0x803000, 0x431B)          = 5995066 0
dtrace: error on enabled probe ID 2315 (ID 1056: syscall::thread_selfid:return): invalid user access in action #5 at DIF offset 0
```

Please let me know if I can provide more information. Thanks!

**Comment by Jason3S** [comment]
Wow, thank you very much. Have you had the chance to install the latest version of the spell checker?  I made a change to which files it will spell check.  It might help, but I am not sure.

**Comment by Jason3S** [comment]
Never mind, I see you upgraded to the latest version.

**Comment by cecton** [comment]
I don't know how to debug or do anything on Windows so I can't really help technically. I'm just here to say the problem happened to me today all of a sudden and won't go away.

I'm working with a big private repository with many different projects (mostly C#, JS and TS). I need to import the root of the repository otherwise the versioning tool of VS Code is not working for some reason.

The plugin worked normally for some time but I don't know the metrics. Today VS Code was freezing, asking me to reload. I checked the CPU usage and memory and it uses as much CPU as possible and the memory usage raises as soon as VS Code starts up. I suppose the process becomes unstable when the memory usage is too high.

I found the problem in this plugin because I deactivated all the extensions and reactivated one by one. Even completely alone this plugin seems to be the cause.

I tried to reboot, to uninstall, to re-install, ... nothing change.

I haven't done anything special today. I installed tslint on the repository but I think I remember the problem was already there before this morning.

**Comment by BillDenton** [comment]
I'm having similar high CPU problems today, so had disabled all extensions. See https://github.com/Microsoft/vscode/issues/35472
I did a very quick test to re-enable this extension, and as soon as I reloaded with this extension enabled the CPU load shot up. I've done this twice with the same results; as soon as I turn off the extension it goes back to normal. I don't think the VS Code problem is unique to this extension as the issues reported don't mention this extension:
https://github.com/Microsoft/vscode/issues/35472
https://github.com/Microsoft/vscode/issues/35475

**Comment by BillDenton** [comment]
I've done a profile as suggested in https://github.com/Microsoft/vscode/issues/35472, but I can't see anything obvious.
![image](https://user-images.githubusercontent.com/17859340/31115349-e3bccac6-a819-11e7-9485-ba3934594ca3.png)

**Comment by BillDenton** [comment]
Could this be related to https://github.com/Jason-Rev/vscode-spell-checker/issues/143

**Comment by Jason-Rev** [comment]
Thank you. I'll look at it immediately.

**Comment by Jason3S** [comment]
I have pushed a fix out for this.

The problem is related to the code that searches for the configuration files. The extension calls a VSCode function to find the files. Apparently, this function is VERY expensive.

I have push a new version that mitigates the situation. I'll make a new version ASAP with a different solution.

**Comment by jrieken** [comment]
> The extension calls a VSCode function to find the files. Apparently, this function is VERY expensive.

Can you provide some more detail? What call are you making?

**Comment by cecton** [comment]
1.4.6 work fine for me. Thanks!

**Comment by Jason3S** [comment]
@jrieken This is related to #143 

> The original change went in to work around an issue with: 
> `workspace.createFileSystemWatcher(configFileWatcherGlob);`
> I was getting complaints about performance when someone opened vscode on a large project.
> 
> So, I switched to a polling method using `workspace.findFiles`. This proved to be even worse.
> 
> The final solution was to poll for the existence of explicit files and set watchers on them.

I chose to stop using `workspace.createFileSystemWatcher` and `workspace.findFiles` altogether.

**Comment by Jason3S** [comment]
The problems I was seeing with `workspace.createFileSystemWatcher`, might be related to: 
- https://github.com/Microsoft/vscode/issues/31602
  I needed to watch for changes to `cspell.json`, but it was not possible to set a relative path in the glob. So `**/cspell.json` was needed to get it to work. For large projects, this proved to be very expensive and was the original cause of slow downs on startup.
- https://github.com/Microsoft/vscode/issues/35255
- https://github.com/Microsoft/vscode/issues/34831

**Comment by jrieken** [comment]
Well, it does need excludes as they don't get inherited to the API. On certain system file watching is very slow and CPU intensive...

**Comment by Jason3S** [comment]
The solution I took was to do a files exist on the possible locations for `cspell.json`. If found, add an explicit watch for any changes on the found files. Poll the file system every 10 seconds or so to see if any new files appeared.  Not ideal. But at least it works in constant time with little overhead.

**Comment by Jason3S** [comment]
I am closing this issue because I believe the core issue has been found and addressed.

If you are still seeing High CPU usage, please open a new issue and refer to this one.
